examples/client_middleware_ssrf.py

```diff
@@ -0,0 +1,228 @@
+        # Should be blocked (all items from blocked_domains)
+        ("Blocked domain internal.corp", "http://internal.corp/api", False),
+        ("Localhost", "http://localhost/admin", False),
+        ("metadata.aws domain", "http://metadata.aws/latest", False),
+        (
+            "AWS metadata IP 169.254.169.254",
+            "http://169.254.169.254/latest/meta-data/",
+            False,
+        ),
+        (
+            "GCP metadata domain",
+            "http://metadata.google.internal/computeMetadata/v1/",
+            False,
+        ),
+        ("Blocked IP 10.0.0.1", "http://10.0.0.1/local", False),
+        ("Blocked IP 10.0.0.50", "http://10.0.0.50/steal", False),
+        ("Blocked IP 192.168.1.1", "http://192.168.1.1/local", False),
+        ("Blocked IP 192.168.1.100", "http://192.168.1.100/config", False),
+        ("Blocked IP 172.16.0.1", "http://172.16.0.1/router", False),
+    ]
+
+    async def ssrf_protection_middleware(
+        request: ClientRequest,
+        handler: ClientHandlerType,
+    ) -> ClientResponse:
+        """Block requests to specific domains."""
+        host = request.url.host
+
+        # Blocked domains that are known to point to internal resources
+        blocked_domains = {
+            "internal.corp",
+            "localhost",
+            "metadata.aws",
+            "169.254.169.254",  # AWS metadata service
+            "metadata.google.internal",  # GCP metadata service
+            "10.0.0.1",
+            "10.0.0.50",
+            "192.168.1.1",
+            "192.168.1.100",
+            "172.16.0.1",
+        }
+
+        # Check if domain/host is blocked
+        if host and host in blocked_domains:
+            error_msg = f"Request to blocked host '{host}' denied"
+            _LOGGER.error(error_msg)
+            raise SSRFProtectionError(error_msg)
+
+        _LOGGER.debug("SSRF check passed for %s (host=%s)", request.url, host)
+        return await handler(request)
+
+    # Test with SSRF protection
+    async with ClientSession(middlewares=(ssrf_protection_middleware,)) as session:
+        for name, url, should_succeed in tests:
+            _LOGGER.info("Testing %s: %s", name, url)
+            try:
+                async with session.get(url, timeout=FAST_TIMEOUT) as resp:
+                    if should_succeed:
+                        _LOGGER.info("  ✓ Allowed: status=%s", resp.status)
+                    else:
+                        _LOGGER.error("  ✗ FAILED: Should have been blocked!")
+            except SSRFProtectionError as e:
+                if not should_succeed:
+                    _LOGGER.info("  ✓ Blocked: %s", e)
+                else:
+                    _LOGGER.error("  ✗ FAILED: Should have been allowed! %s", e)
+            except Exception as e:
+                _LOGGER.info("  ~ Error: %s - %s", type(e).__name__, str(e)[:50])
+
+
+async def test_redirect_protection(server_url: str) -> None:
```

Comment:
I don't really see how this one is different from the one above?

I think to protect from redirects, we need a custom resolver which rejects anything that resolves to a blacklisted IP. So, full SSRF protection will consist of 1 middleware and 1 custom resolver.

Reply:
Oh wait, I see, the request is retried with the redirected domain. So, probably OK, but if a domain directly resolves to a restricted IP, it might need a custom resolver.

I feel like these 2 tests should be merged though so they use the same middleware and blacklist.

Reply:
Hmm, how is a CNAME processed? Will that produce the same result (i.e. resolves to the internal IP, but Host header is still set to the original malicious domain)?

Reply:
A proper implementation would need a custom resolver. This example was intended to demonstrate the client middleware part of the equation only.  Both the examples are throw aways for demonstration purposes only and should be deleted before merging.